lib/python/operations.cpp

```diff
@@ -82,6 +83,65 @@ void bind_midpoints(py::module &m) {
   });
 }
 
+std::tuple<Variable, std::optional<Coords>>
+extract_where_argument(const py::object &arg) {
+  if (py::isinstance<Variable>(arg)) {
+    return {arg.cast<Variable>(), std::nullopt};
+  }
+  auto da = arg.cast<DataArray>();
+  if (!da.masks().empty()) {
+    throw std::invalid_argument("Arguments of 'where' must not have masks");
+  }
+  return {da.data(), std::optional(da.coords())};
+}
+
+Variable apply_where(const Variable &c_data, const Variable &x_data,
+                     const Variable &y_data,
+                     const std::optional<Coords> &c_coords,
+                     const std::optional<Coords> &x_coords,
+                     const std::optional<Coords> &y_coords) {
+  py::gil_scoped_release _release;
+
+  if (c_coords.has_value() && x_coords.has_value()) {
+    dataset::expect::coords_are_superset(c_coords.value(), x_coords.value(),
+                                         "where (condition and x)");
+  }
+  if (c_coords.has_value() && y_coords.has_value()) {
+    dataset::expect::coords_are_superset(c_coords.value(), y_coords.value(),
+                                         "where (condition and y)");
+  }
+  if (x_coords.has_value() && y_coords.has_value()) {
+    if (x_coords.value() != y_coords.value()) {
+      throw except::CoordMismatchError(
+          "Expected coords of x and y to match in 'where' operation");
+    }
+  }
+
+  return where(c_data, x_data, y_data);
+}
+
+void bind_where(py::module &m) {
+  m.def(
+      "where",
+      [](const py::object &condition, const py::object &x,
+         const py::object &y) {
+        auto [c_data, c_coords] = extract_where_argument(condition);
+        auto [x_data, x_coords] = extract_where_argument(x);
+        auto [y_data, y_coords] = extract_where_argument(y);
+        const Variable new_data =
+            apply_where(c_data, x_data, y_data, c_coords, x_coords, y_coords);
+
+        if (x_coords.has_value()) {
+          return py::cast(DataArray(new_data, x_coords.value(), {}));
```

Comment:
Maybe I missed something, but it is not clear to me why we require c_coords to be a superset, but then drop those coords and use only value coords?

Reply:
I'm not sure about this either. I thought that we want to make sure that the condition's coords match the data. But it seemed too restrictive to require completely identical coords.  Did I get it the wrong way around? Should the data's coords be a superset of the condition's?

Reply:
Could we just return the union? Or require that the condition is a variable?

Reply:
> Could we just return the union?

I guess we could. It's not exactly cheap to do a union of 3 inputs but it can be done.

> Or require that the condition is a variable?

I wouldn't do that because it encourages discarding coords from the inputs when building the condition.

Reply:
> > Could we just return the union?
> 
> I guess we could. It's not exactly cheap to do a union of 3 inputs but it can be done.

I thought we should require that value coords are the same, but return the union of condition and value coords? Similar to how regular binary ops do it (there should be helpers for that somewhere).